[Debt] Migrate codegen to `client-preset` (#8673)
GCTC-NTGC/gc-digital-talent: documentation/graphq.md

### 🧩 Fragments

The preset allows us to better utilize fragments. This means we can compose our queries and build them "bottom up" similar to the react data flow, but in reverse. Instead of writing one large query and pulling out the pieces we need for each component, we can use the component to tell us what it needs and build the query from that.

### 🗺️ Collocation

Currently, we achieve this by placing our `*.operations.graphql` files as close as possible to the page/component that uses them. Now, we can (and should) place them in the file that uses them. No more hunting for the proper file!

### 👺 Masking

With the fragments, we can use what is referred to as "fragment masking". What this provides is a way to strongly type the specific fragment used in a query for our consumption. Previously, we had some problems where we added additional information to a component, but since our type system was very relaxed (it showed us what *could* be queried, not what *was* queried) we sometimes forgot to update the query and ended up with empty values.

Now, we get a type error if we use data we never queried and will get a build error.

## 🛠️ Implementation

> All examples are related and show how we would build a page and its operations bottom up.

Reviewer: ```suggestion
All examples are related and show how to build a page and its operations from the bottom up.
```

### ⚛️ Component

> When defining a component that needs to consume data from the API, you need to do a few things for the code generator to work.

Reviewer: ```suggestion
When defining a component that needs to consume data from the API, there are a few things to do for the code generator to work.
```

 1. Import the `graphql` helper which accepts a graphql document string (for components this will be a fragment)
 2. Add the fragment to the prop types so it can be passed in
 3. Mask the fragment for strong typing of the fields queried

```tsx
// PoolCard.tsx
import { FragmentType, getFragment, graphql } from "@gc-digital-talent/graphql";

export const PoolCard_PoolFragment = graphql(/* GraphQL */ `
  fragment PoolCard_PoolFragment on Pool {
    id
    name {
      en
      fr
    }
  }
`)

type PoolCardProps = {
  pool: FragmentType<typeof PoolCard_PoolFragment>;
}

const PoolCard = ({ pool: poolFragment }: PoolCardProps) => {
  // This extracts and strongly types the fields we defined in the fragment
  const pool = getFragment(PoolCard_PoolFragment, poolFragment);

  return (
    <>
      <p>{getLocalizedName(pool.name, intl)}</p>
      {/** ⚠️ This will display an error and fail to build since we never queried it in the fragment! */}
      <p>{pool.stream}<p>
    </>
  );
}
```

### 📖 Storybook

> Since our props require a specific type that comes from the code generator, we need to massage the mock data we pass into stories. Thankfully, we have a nice little helper to do this.

Author: Sounds good! Hopefully this makes sense: [add reference to fragment helper](https://github.com/GCTC-NTGC/gc-digital-talent/pull/8673/commits/e8bafb76c1db0ff2a230a7d45862453fca3e74cd)

```tsx
// PoolCard.stories.tsx
import { makeFragmentData } from "@gc-digital-talent/graphql";
import { fakePools } from "@gc-digital-talent/fake-data";

import PoolCard, { PoolCard__PoolFragment } from "./PoolCard";

const mockPools = fakePools();
const mockPoolFragment = makeFragmentData(mockPools[0], PoolCard__PoolFragment);

const Template = () => <PoolCard pool={mockPoolFragment} />;
```

### 🧪 Testing

> Similar to storybook, we just need to create the fragment using the helper.

Author: Added in [add reference to fragment helper](https://github.com/GCTC-NTGC/gc-digital-talent/pull/8673/commits/e8bafb76c1db0ff2a230a7d45862453fca3e74cd)
